### PR TITLE
[isal] Update port

### PR DESCRIPTION
**Pull request by anders-wind**
**Describe the pull request**
Trying to get isal to support dynamic linking as described in usage

- #### What does your PR fix?
I add a cmake config target for isal as described in the usage file.

- #### Which triplets are supported/not supported? Have you updated the [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt)?
all / No

- #### Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)?
Hopefully :) 

- #### If you have added/updated a port: Have you run `./vcpkg x-add-version --all` and committed the result?
I have 

**If you are still working on the PR, open it as a Draft: https://github.blog/2019-02-14-introducing-draft-pull-requests/**

### CODE REVIEW

**Comment by anders-wind** [comment]
When I look at the ouitput from the CI pipeline it doesnt look like its copying over an `isal.so` - However I dont know if thats just because its only testing the static library setup

**Comment by anders-wind** [comment]
That was better - now libisal.so is in the output.

**Comment by dg0yt** [comment]
Both `CMakeLists.txt` and `portfile.cmake` look very strange, with much pointless code. If not using upstream's build systems, the port's vendored build system should be modified to a more standard Modern CMake style, directly exporting CMake config to namespace `unofficial::<pkg>::`. Everything else is just adding more technical debt.

**Comment by anders-wind** [comment]
Sadly I don't have the time to completely rewrite the cmake system and currently I cannot use isal because the shared library is not exposed.

**Comment by dg0yt** [comment]
> and currently I cannot use isal because the shared library is not exposed

Shared linkage on Linux is untested in vcpkg CI (which is really unfortunate). 

>the static library is not compiled with position independent code enabled.

This is because this is such a poor port... this must be fixed, too.

> Sadly I don't have the time to completely rewrite the cmake system

That's okay. Ideally, this task would be taken over by one of the vcpkg maintainers. The `CMakeList.txt` is pointless. It just runs `nmake` or `make`. `portfile.cmake` should use proper maintainer functions, maybe using the package's autotools build system for all platforms.

**Comment by FrankXie05** [comment]
This port is too old, maybe it should be rewritten. 🤔

**Comment by anders-wind** [comment]
So this PR wont be merged?

**Comment by FrankXie05** [comment]
@anders-wind Yes, sorry, since this port is ugly and old now, the best way to do is to rewrite it. 
So this PR you can close it, or try to convert it to a draft to rewrite it, you can ask anyone for help, and I'm sure they will be happy to help you. :)

**Comment by JackBoosY** [comment]
I will take over this PR.

**Comment by JackBoosY** [comment]
Ping for response.

**Comment by anders-wind** [comment]
Ill try it out now :)

**Comment by anders-wind** [comment]
So I tried to change our project's vcpkg folder to my forks and checkout out the patch-3 branch. 

Then I can cmake and got the following output
```
[build] Installing 125/135 isal:x64-linux...
[build] Building isal[core]:x64-linux...
[build] -- Installing port from location: /home/anderswind/projects/xxxxx/vcpkg/buildtrees/versioning_/versions/isal/7fb5582ae162e9533b75b50c629edd88cecc58ad
[build] -- Downloading https://github.com/01org/isa-l/archive/v2.25.0.tar.gz -> 01org-isa-l-v2.25.0.tar.gz...
[build] -- Extracting source /home/anderswind/projects/xxxxx/vcpkg/downloads/01org-isa-l-v2.25.0.tar.gz
[build] -- Applying patch fix-nmake.patch
[build] -- Using source at /home/anderswind/projects/xxxxx/vcpkg/buildtrees/isal/src/v2.25.0-2318393b30.clean
[build] -- Getting CMake variables for x64-linux-dbg
[build] -- Getting CMake variables for x64-linux-rel
[build] -- Generating configure for x64-linux
[build] -- Finished generating configure for x64-linux
[build] -- Configuring x64-linux-dbg
[build] -- Configuring x64-linux-rel
[build] -- Building x64-linux-dbg
[build] -- Installing x64-linux-dbg
[build] -- Building x64-linux-rel
[build] -- Installing x64-linux-rel
[build] -- Installing: /home/anderswind/projects/xxxxx/vcpkg/packages/isal_x64-linux/share/isal/usage
[build] -- Installing: /home/anderswind/projects/xxxxx/vcpkg/packages/isal_x64-linux/share/isal/copyright
[build] -- Performing post-build validation
[build] -- Performing post-build validation done
...
[build] The package isal is compatible with built-in CMake targets:
[build] 
[build]     find_package(isal CONFIG REQUIRED)
[build]     target_link_libraries(main PRIVATE ISAL::isa-l) for static linkage or
[build]     target_link_libraries(main PRIVATE ISAL::isal) for dynamic linkage
...
[build] CMake Error at vcpkg/scripts/buildsystems/vcpkg.cmake:573 (_add_executable):
[build]   Target "RunBenchmark" links to target "ISAL::isal" but the target was not
[build]   found.  Perhaps a find_package() call is missing for an IMPORTED target, or
[build]   an ALIAS target is missing?
```
So I changed the dependency name to `ISAL::isa-l`
For which i get:
```
[build] -- Configuring done
[build] CMake Error in xxxxx/CMakeLists.txt:
[build]   Imported target "ISAL::isa-l" includes non-existent path
[build] 
[build]     "/home/anderswind/projects/xxxxx/build/vcpkg_installed/x64-linux/include/isal-l"
[build] 
[build]   in its INTERFACE_INCLUDE_DIRECTORIES.  Possible reasons include:
[build] 
[build]   * The path was deleted, renamed, or moved to another location.
[build] 
[build]   * An install or uninstall procedure did not complete successfully.
[build] 
[build]   * The installation package was faulty and references files it does not
[build]   provide.
```

I might have done something wrong?

**Comment by anders-wind** [comment]
```
total 1.4M
drwxrwxr-x 21 anderswind anderswind 4.0K Jun 27 12:17 absl
-rw-r--r--  1 anderswind anderswind  48K Jun 27 12:17 amqp_framing.h
-rw-r--r--  1 anderswind anderswind  94K Jun 27 12:17 amqp.h
-rw-r--r--  1 anderswind anderswind 9.3K Jun 27 12:17 amqp_ssl_socket.h
-rw-r--r--  1 anderswind anderswind 2.2K Jun 27 12:17 amqp_tcp_socket.h
drwxrwxr-x 14 anderswind anderswind 4.0K Jun 27 12:17 arrow
drwxrwxr-x 18 anderswind anderswind 4.0K Jun 27 12:17 aws
drwxrwxr-x  2 anderswind anderswind 4.0K Jun 27 12:17 benchmark
drwxrwxr-x 59 anderswind anderswind 4.0K Jun 27 12:17 boost
drwxrwxr-x  2 anderswind anderswind 4.0K Jun 27 12:17 brotli
-rw-r--r--  1 anderswind anderswind 6.1K Jun 27 12:17 bzlib.h
drwxrwxr-x  3 anderswind anderswind 4.0K Jun 27 12:17 cnpy
drwxrwxr-x  2 anderswind anderswind 4.0K Jun 27 12:17 curl
drwxrwxr-x  2 anderswind anderswind 4.0K Jun 27 12:17 cute
-rw-r--r--  1 anderswind anderswind  56K Jun 27 12:17 cxxopts.hpp
-rw-r--r--  1 anderswind anderswind 2.0K Jun 27 12:17 evdns.h
drwxrwxr-x  2 anderswind anderswind 4.0K Jun 27 12:17 event2
-rw-r--r--  1 anderswind anderswind 2.7K Jun 27 12:17 event.h
-rw-r--r--  1 anderswind anderswind 2.0K Jun 27 12:17 evhttp.h
-rw-r--r--  1 anderswind anderswind 2.0K Jun 27 12:17 evrpc.h
-rw-r--r--  1 anderswind anderswind 1.8K Jun 27 12:17 evutil.h
drwxrwxr-x  2 anderswind anderswind 4.0K Jun 27 12:17 fmt
drwxrwxr-x  3 anderswind anderswind 4.0K Jun 27 12:17 frozen
drwxrwxr-x  2 anderswind anderswind 4.0K Jun 27 12:17 gflags
drwxrwxr-x  2 anderswind anderswind 4.0K Jun 27 12:17 glog
drwxrwxr-x  3 anderswind anderswind 4.0K Jun 27 12:17 gmock
drwxrwxr-x  3 anderswind anderswind 4.0K Jun 27 12:17 gtest
-rw-rw-r--  1 anderswind anderswind 259K Jun 27 12:17 httplib.h
drwxrwxr-x  2 anderswind anderswind 4.0K Jun 27 12:17 isa-l
-rw-r--r--  1 anderswind anderswind  590 Jun 27 12:17 isa-l.h
drwxrwxr-x  2 anderswind anderswind 4.0K Jun 27 12:17 libpq
-rw-r--r--  1 anderswind anderswind 2.2K Jun 27 12:17 libpq-events.h
-rw-r--r--  1 anderswind anderswind  25K Jun 27 12:17 libpq-fe.h
-rw-r--r--  1 anderswind anderswind  29K Jun 27 12:17 lz4frame.h
-rw-r--r--  1 anderswind anderswind 2.0K Jun 27 12:17 lz4frame_static.h
-rw-r--r--  1 anderswind anderswind  40K Jun 27 12:17 lz4.h
-rw-r--r--  1 anderswind anderswind  20K Jun 27 12:17 lz4hc.h
-rw-r--r--  1 anderswind anderswind 3.7K Jun 27 12:17 magic_enum_fuse.hpp
-rw-r--r--  1 anderswind anderswind  56K Jun 27 12:17 magic_enum.hpp
-rw-r--r--  1 anderswind anderswind  28K Jun 27 12:17 mimalloc.h
-rw-r--r--  1 anderswind anderswind 3.4K Jun 27 12:17 mimalloc-new-delete.h
-rw-r--r--  1 anderswind anderswind 3.1K Jun 27 12:17 mimalloc-override.h
drwxrwxr-x  4 anderswind anderswind 4.0K Jun 27 12:17 nlohmann
drwxrwxr-x  2 anderswind anderswind 4.0K Jun 27 12:17 openssl
drwxrwxr-x  5 anderswind anderswind 4.0K Jun 27 12:17 parquet
-rw-r--r--  1 anderswind anderswind  323 Jun 27 12:17 pg_config_ext.h
-rw-r--r--  1 anderswind anderswind  32K Jun 27 12:17 pg_config.h
-rw-r--r--  1 anderswind anderswind  15K Jun 27 12:17 pg_config_manual.h
-rw-r--r--  1 anderswind anderswind 1.1K Jun 27 12:17 pg_config_os.h
-rw-r--r--  1 anderswind anderswind 2.2K Jun 27 12:17 postgres_ext.h
drwxrwxr-x  4 anderswind anderswind 4.0K Jun 27 12:17 postgresql
-rw-r--r--  1 anderswind anderswind 2.9K Jun 27 12:17 pugiconfig.hpp
-rw-r--r--  1 anderswind anderswind  53K Jun 27 12:17 pugixml.hpp
drwxrwxr-x  5 anderswind anderswind 4.0K Jun 27 12:17 rapidjson
drwxrwxr-x  2 anderswind anderswind 4.0K Jun 27 12:17 re2
-rw-r--r--  1 anderswind anderswind  65K Jun 27 12:17 s2n.h
drwxrwxr-x  3 anderswind anderswind 4.0K Jun 27 12:17 SimpleAmqpClient
-rw-r--r--  1 anderswind anderswind 5.3K Jun 27 12:17 snappy-c.h
-rw-r--r--  1 anderswind anderswind 9.7K Jun 27 12:17 snappy.h
-rw-r--r--  1 anderswind anderswind 7.1K Jun 27 12:17 snappy-sinksource.h
-rw-r--r--  1 anderswind anderswind 2.6K Jun 27 12:17 snappy-stubs-public.h
drwxrwxr-x  6 anderswind anderswind 4.0K Jun 27 12:17 spdlog
drwxrwxr-x  3 anderswind anderswind 4.0K Jun 27 12:17 tao
drwxrwxr-x 10 anderswind anderswind 4.0K Jun 27 12:17 thrift
-rw-r--r--  1 anderswind anderswind  31K Jun 27 12:17 utf8proc.h
drwxrwxr-x 11 anderswind anderswind 4.0K Jun 27 12:17 xercesc
-rw-r--r--  1 anderswind anderswind  17K Jun 27 12:17 zconf.h
-rw-r--r--  1 anderswind anderswind  26K Jun 27 12:17 zdict.h
-rw-r--r--  1 anderswind anderswind  96K Jun 27 12:17 zlib.h
-rw-r--r--  1 anderswind anderswind 3.8K Jun 27 12:17 zstd_errors.h
-rw-r--r--  1 anderswind anderswind 146K Jun 27 12:17 zstd.h
```

**Comment by anders-wind** [comment]
So thats a bit weird

**Comment by JackBoosY** [comment]
`isa-l`, not `isal-1`.

**Comment by JackBoosY** [comment]
@anders-wind Can you please test again?

Thanks.

**Comment by JackBoosY** [comment]
Sorry, it should be okay now.

**Comment by anders-wind** [comment]
It works 🎉  Thank you so much - further more it seems FPIC is now correctly applied so a bunch of my issues dissapeared. 
My only remaining question would be concerning the dynamic library linking. The documentation says:
```
[build] The package isal is compatible with built-in CMake targets:
[build] 
[build]     find_package(isal CONFIG REQUIRED)
[build]     target_link_libraries(main PRIVATE ISAL::isa-l) for static linkage or
[build]     target_link_libraries(main PRIVATE ISAL::isal) for dynamic linkage
```
But ISAL::isal is not an existing target ` Target "XXXXX" links to target "ISAL::isal" but the target was not found.`, and I also cannot find libisal.so in `vcpkg_installed/x64-linux/lib/` so I guess its not compiled?. 
Will it be possible to support adding it as a dynamic library or should we just fix the documentation? 
With the added FPIC I am not in as much need of the dynamic library so its up to you :)

**Comment by JackBoosY** [comment]
@anders-wind Please note `x64-linux` / `x64-osx` is a static triplet, so you need to use target `ISAL::isa-l`.

**Comment by dg0yt** [comment]
> My only remaining question would be concerning the dynamic library linking. The documentation says:
> 
> ```
> [build] The package isal is compatible with built-in CMake targets:
> [build] 
> [build]     find_package(isal CONFIG REQUIRED)
> [build]     target_link_libraries(main PRIVATE ISAL::isa-l) for static linkage or
> [build]     target_link_libraries(main PRIVATE ISAL::isal) for dynamic linkage
> ```
> 
> But ISAL::isal is not an existing target ` Target "XXXXX" links to target "ISAL::isal" but the target was not found.`, and I also cannot find libisal.so in `vcpkg_installed/x64-linux/lib/` so I guess its not compiled?. Will it be possible to support adding it as a dynamic library or should we just fix the documentation? 

The documentation should be changed similar to zstd:
~~~cmake
...
target_link_libraries(main PRIVATE $<IF:$<TARGET_EXISTS:ISAL::isa-l>,ISAL::isa-l,ISAL::isal>)
~~~

**Comment by JackBoosY** [comment]
@anders-wind Any other questions?

**Comment by anders-wind** [comment]
> @anders-wind Any other questions?

Nope nothing specifically to this port. Thanks for the help! 🎉

**Comment by JackBoosY** [comment]
> > My only remaining question would be concerning the dynamic library linking. The documentation says:
> > ```
> > [build] The package isal is compatible with built-in CMake targets:
> > [build] 
> > [build]     find_package(isal CONFIG REQUIRED)
> > [build]     target_link_libraries(main PRIVATE ISAL::isa-l) for static linkage or
> > [build]     target_link_libraries(main PRIVATE ISAL::isal) for dynamic linkage
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > But ISAL::isal is not an existing target ` Target "XXXXX" links to target "ISAL::isal" but the target was not found.`, and I also cannot find libisal.so in `vcpkg_installed/x64-linux/lib/` so I guess its not compiled?. Will it be possible to support adding it as a dynamic library or should we just fix the documentation?
> 
> The documentation should be changed similar to zstd:
> 
> ```cmake
> ...
> target_link_libraries(main PRIVATE $<IF:$<TARGET_EXISTS:ISAL::isa-l>,ISAL::isa-l,ISAL::isal>)
> ```

I will do that tomorrow.

**Review by JackBoosY — DISMISSED** [review]

File: ports/isal/isalConfig.cmake [hunk 1]
```diff
@@ -1,17 +1,73 @@
...
+    if (arg_DEBUG_DYNAMIC)
+        set(ISAL_PROPERTIES IMPORTED_LOCATION_DEBUG "${arg_DEBUG_DYNAMIC}" IMPORTED_IMPLIB_DEBUG "${arg_DEBUG_STATIC}")
+    else()
+        set(ISAL_PROPERTIES IMPORTED_LOCATION_DEBUG "${arg_DEBUG_STATIC}")
+    endif()
+    
+    if (arg_RELEASE_DYNAMIC)
+        set(ISAL_PROPERTIES IMPORTED_LOCATION_DEBUG "${arg_RELEASE_DYNAMIC}" IMPORTED_IMPLIB_DEBUG "${arg_RELEASE_STATIC}")
+    else()
+        set(ISAL_PROPERTIES IMPORTED_LOCATION_DEBUG "${arg_RELEASE_STATIC}")
```
Comment (LilyWangLL) [inline ports/isal/isalConfig.cmake:14]:
```suggestion
        set(ISAL_PROPERTIES IMPORTED_LOCATION_RELEASE "${arg_RELEASE_DYNAMIC}" IMPORTED_IMPLIB_RELEASE "${arg_RELEASE_STATIC}")
    else()
        set(ISAL_PROPERTIES IMPORTED_LOCATION_RELEASE "${arg_RELEASE_STATIC}")
```

File: ports/isal/isalConfig.cmake [hunk 2]
```diff
@@ -1,17 +1,73 @@
...
+            RELEASE_STATIC "${_IMPORT_PREFIX}/lib/libisal.a"
+            INCLUDE_DIR "${_IMPORT_PREFIX}/include"
+            TYPE STATIC
+        )
+    else()
+        set_library_target(
+            NAMESPACE "ISAL"
+            LIB_NAME "isal"
+            DEBUG_STATIC "${_IMPORT_PREFIX}/debug/lib/libisal.so"
+            RELEASE_STATIC "${_IMPORT_PREFIX}/lib/libisal.so"
```
Comment (LilyWangLL) [inline ports/isal/isalConfig.cmake:68]:
```suggestion
            DEBUG_DYNAMIC "${_IMPORT_PREFIX}/debug/lib/libisal.so"
            RELEASE_DYNAMIC "${_IMPORT_PREFIX}/lib/libisal.so"
```

File: ports/isal/portfile.cmake [hunk 3]
```diff
@@ -1,46 +1,74 @@
...
+
+    if (NOT DEFINED VCPKG_BUILD_TYPE OR VCPKG_BUILD_TYPE STREQUAL "debug")
+        set(NMAKE_BINARY_DIR "${CURRENT_BUILDTREES_DIR}/${TARGET_TRIPLET}-dbg")
+        if (VCPKG_LIBRARY_LINKAGE STREQUAL "dynamic")
+            file(INSTALL "${NMAKE_BINARY_DIR}/isa-l.dll" DESTINATION "${CURRENT_PACKAGES_DIR}/debug/bin")
+            file(INSTALL "${NMAKE_BINARY_DIR}/isa-l.lib" DESTINATION "${CURRENT_PACKAGES_DIR}/debug/lib")
+        else()
+            file(INSTALL "${NMAKE_BINARY_DIR}/isa-l_static.lib" DESTINATION "${CURRENT_PACKAGES_DIR}/debug/lib")
+        endif()
+
```
Comment (LilyWangLL) [inline ports/isal/portfile.cmake:46]:
```suggestion
```

**Review by LilyWangLL — CHANGES_REQUESTED** [review]

**Comment by JackBoosY** [comment]
@anders-wind Sorry, can you please test again?

Thanks.

**Comment by anders-wind** [comment]
I'm on festival so I won't be able to test it before Monday unfortunately, sorry

**Comment by anders-wind** [comment]
Just tried it out :) still works flawlessly on `c2b9e34be` 
The cmake example is not also update to:
```
[cmake] The package isal is compatible with built-in CMake targets:
[cmake] 
[cmake]     find_package(isal CONFIG REQUIRED)
[cmake]     target_link_libraries(main PRIVATE $<IF:$<TARGET_EXISTS:ISAL::isa-l>,ISAL::isa-l,ISAL::isal>)
```
I was wondering if the include path will be the same for the dynamic library (currenlty `<isa-l/...h>`)? Just wondering because I dont know if the `ISA-L` vs `ISAL` has any special meaning.

**Comment by dg0yt** [comment]
> I was wondering if the include path will be the same for the dynamic library (currenlty `<isa-l/...h>`)? Just wondering because I dont know if the `ISA-L` vs `ISAL` has any special meaning.

The include path should be the same, and it is to be carried by the targets. However, for windows, the content usally isn't 100% the same because for dynamic linkage, symbols needs special markup.
(Different names for static vs. shared lib usually come from upstream wanting to build both at once, similar to autotools projects. Not used in vcpgk.)

**Review by LilyWangLL — DISMISSED** [review]

File: ports/isal/portfile.cmake [hunk 4]
```diff
@@ -1,46 +1,73 @@
...
-        REPO 01org/isa-l
-        REF v2.25.0
-        SHA512 aa556c8ba26b4637493b3de50a23636668bcfd71249029c52fe6983d0bcf120d1b91f39aaa259cb58e59448d401366f3bfaaee24609db7e6a1cd3fdf1a953efe
-        HEAD_REF master
-)
+if (VCPKG_TARGET_ARCHITECTURE STREQUAL "x86" OR VCPKG_TARGET_ARCHITECTURE STREQUAL "arm")
+    message(FATAL_ERROR "Intel(R) Intelligent Storage Acceleration Library currently only supports x64 architecture")
+elseif (NOT VCPKG_TARGET_ARCHITECTURE STREQUAL "x64")
+    message(FATAL_ERROR "Unsupported architecture: ${VCPKG_TARGET_ARCHITECTURE}")
+endif ()
```
Comment (Thomas1664) [inline ports/isal/portfile.cmake:5]: Not needed, already handled by `supports` in `vcpkg.json`

**Comment by vicroms** [comment]
Thanks @JackBoosY for taking over the PR and everyone else for their reviews!

**Pull request merged**